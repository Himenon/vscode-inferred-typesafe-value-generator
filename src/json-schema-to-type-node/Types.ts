import type { JSONSchema7, JSONSchema7TypeName } from "json-schema";

export type Schema = JSONSchema7;

export type JSONSchemaTypeName = JSONSchema7TypeName;

export type JSONSchemaDefinition = Schema | boolean;

export interface Reference {
  $ref: string;
  summary?: string;
  description?: string;
}

export interface UnSupportSchema extends Omit<Schema, "type"> {
  type: JSONSchemaTypeName[];
}

export interface OneOfSchema extends Omit<Schema, "oneOf"> {
  oneOf: Schema[];
}

export interface AllOfSchema extends Omit<Schema, "allOf"> {
  allOf: Schema[];
}

export interface AnyOfSchema extends Omit<Schema, "anyOf"> {
  anyOf: Schema[];
}

export interface MultiTypeSchema extends Omit<Schema, "type"> {
  type: JSONSchemaTypeName[];
}

export interface ObjectSchema extends Omit<Schema, "type"> {
  type: "object";
}

export interface FunctionSchema extends Omit<ObjectSchema, "format"> {
  format: "function";
}

export interface ObjectSchemaWithAdditionalProperties extends ObjectSchema {
  additionalProperties: JSONSchemaDefinition;
}

export interface ArraySchema extends Omit<Schema, "type"> {
  type: "array";
}

export interface PrimitiveSchema extends Omit<Schema, "type"> {
  type: "string" | "number" | "integer" | "boolean" | "null";
}

export interface AnySchema extends Omit<Schema, "type"> {
  type: "any";
}
