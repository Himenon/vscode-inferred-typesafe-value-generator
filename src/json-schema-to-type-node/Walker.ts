import * as TypeGuard from "./TypeGuard";
import type * as Types from "./Types";
import * as DotProp from "dot-prop";
import * as Logger from "../logger";

export interface ConvertedObject {
  value: any;
  errorMessages: string[];
}

export class Walker {
  private readonly schema: Types.Schema;
  constructor(schema: Types.Schema) {
    this.schema = JSON.parse(JSON.stringify(schema));
  }

  private _convert = (schema?: Types.Schema): any => {
    if (!schema) {
      return "undefined";
    }
    if (TypeGuard.isReference(schema)) {
      return this.reference(schema);
    }
    if (TypeGuard.isObjectSchema(schema)) {
      if (TypeGuard.isFunctionSchema(schema)) {
        return this.functionTypeValue(schema);
      }
      return this.convertObject(schema);
    }
    if (TypeGuard.isPrimitiveSchema(schema)) {
      return this.primitive(schema);
    }
    if (TypeGuard.isArraySchema(schema)) {
      return this.array(schema);
    }
    if (TypeGuard.isMultiTypeSchema(schema)) {
      return this.multiTypeValue(schema);
    }
    if (TypeGuard.isAnyOfSchema(schema)) {
      return this.anyOfValue(schema);
    }
    if (TypeGuard.isOneOfSchema(schema)) {
      return this.oneOfValue(schema);
    }
    if (TypeGuard.isAllOfSchema(schema)) {
      return this.allOfValue(schema);
    }
    throw new Error(`UnSupport JSON Schema: ${JSON.stringify(schema)}`);
  };

  public convert = (schema?: Types.Schema): ConvertedObject => {
    try {
      const value = this._convert(schema);
      return {
        value,
        errorMessages: [],
      };
    } catch (error) {
      const e = error as Error;
      Logger.error(e.stack);
      return {
        value: undefined,
        errorMessages: [e.message],
      };
    }
  };

  private undefinedValue = (): any => {
    return undefined;
  };

  private reference = (schema: Types.Reference): any => {
    const decodeRef = decodeURIComponent(schema.$ref);
    const ref = decodeRef.replace(/\./g, "\\.").replace(/^#\//, "").replace(/\//g, ".");
    const refSchema = DotProp.get(this.schema, ref) as any;
    return this._convert(refSchema);
  };

  private array = (schema: Types.ArraySchema): any[] => {
    const items = schema.items;
    if (!items || items === true) {
      return [];
    }
    if (!Array.isArray(items)) {
      return [this._convert(items)];
    }
    return items.map((item) => {
      if (typeof item === "boolean") {
        return "";
      }
      return this._convert(item);
    });
  };

  private allOfValue = (schema: Types.AllOfSchema) => {
    if (schema.allOf.length > 0) {
      const allOfObject = schema.allOf.reduce((previous, current) => {
        const nextSchema = this._convert(current);
        return { ...previous, ...nextSchema };
      }, {});
      return allOfObject;
    }
    return this.undefinedValue();
  };

  private oneOfValue = (schema: Types.OneOfSchema) => {
    if (schema.oneOf.length > 0) {
      return this._convert(schema.oneOf[0]);
    }
    return this.undefinedValue();
  };

  private anyOfValue = (schema: Types.AnyOfSchema) => {
    if (schema.anyOf.length > 0) {
      return this._convert(schema.anyOf[0]);
    }
    return this.undefinedValue();
  };

  private multiTypeValue = (schema: Types.MultiTypeSchema) => {
    const types = schema.type;
    if (types.length > 0) {
      return this._convert({ type: types[0] });
    }
    return this.undefinedValue();
  };

  private stringValue = (schema: Types.PrimitiveSchema): string => {
    let estimateValue: any;
    if (schema.const) {
      estimateValue = schema.const;
    } else if (schema.default) {
      estimateValue = schema.default;
    } else if (schema.enum && schema.enum.length > 0) {
      estimateValue = schema.enum[0];
    }
    if (typeof estimateValue === "string") {
      return estimateValue;
    }
    return "";
  };

  private numberValue = (schema: Types.PrimitiveSchema): number => {
    let estimateValue: any;
    if (schema.const) {
      estimateValue = schema.const;
    } else if (schema.default) {
      estimateValue = schema.default;
    } else if (schema.enum && schema.enum.length > 0) {
      estimateValue = schema.enum[0];
    }
    if (estimateValue) {
      if (typeof estimateValue === "string") {
        const value = parseInt(estimateValue, 10);
        return value || 0;
      } else if (typeof estimateValue === "number") {
        return estimateValue;
      }
    }
    return 0;
  };

  private booleanValue = (schema: Types.PrimitiveSchema): boolean => {
    let estimateValue: any;
    if (schema.default) {
      estimateValue = schema.default;
    } else if (schema.enum && schema.enum.length > 0) {
      estimateValue = schema.enum[0];
    }
    if (typeof estimateValue === "boolean") {
      return estimateValue;
    }
    if (typeof estimateValue === "string") {
      return Boolean(estimateValue);
    }
    return false;
  };

  private primitive = (schema: Types.PrimitiveSchema) => {
    switch (schema.type) {
      case "boolean":
        return this.booleanValue(schema);
      case "integer":
      case "number":
        return this.numberValue(schema);
      case "string":
        return this.stringValue(schema);
      default:
        throw new Error("UNSupport Type!");
    }
  };

  private convertObject = (schema: Types.ObjectSchema) => {
    const properties = schema.properties || {};
    const obj: Record<string, any> = {};
    Object.entries(properties).forEach(([key, value]) => {
      if (typeof value === "boolean") {
        return;
      }
      obj[key] = this._convert(value);
    });
    return obj;
  };

  /**
   * TODO
   */
  private functionTypeValue = (_schema: Types.FunctionSchema) => {
    return undefined;
  };
}
