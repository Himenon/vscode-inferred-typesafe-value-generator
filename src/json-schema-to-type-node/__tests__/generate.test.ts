import { Schema } from "../Types";
import * as Module from "../Walker";

const sampleJsonSchema = {
  $schema: "http://json-schema.org/draft-07/schema#",
  $ref: "#/definitions/Sample.Config",
  definitions: {
    "Sample.Config": {
      type: "object",
      properties: { name: { type: "string" }, value: { type: "string" } },
      required: ["value"],
      additionalProperties: false,
    },
  },
} as Schema;

describe("dot property", () => {
  test("sampleJsonSchema", () => {
    const walker = new Module.Walker(sampleJsonSchema);
    const { value } = walker.convert(sampleJsonSchema);
    expect(value).toMatchSnapshot();
  });
});
