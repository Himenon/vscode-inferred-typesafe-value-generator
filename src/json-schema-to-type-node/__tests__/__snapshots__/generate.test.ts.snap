// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dot property sampleJsonSchema 1`] = `
Object {
  "name": "",
  "value": "",
}
`;
